qiime2/plugin/plugin.py

```diff
@@ -248,22 +251,77 @@ def register_semantic_types(self, *type_fragments):
                 SemanticTypeFragmentRecord(
                     fragment=type_fragment, plugin=self)
 
-    def register_semantic_type_to_format(self, semantic_type, artifact_format):
-        if not issubclass(artifact_format, DirectoryFormat):
-            raise TypeError("%r is not a directory format." % artifact_format)
+    def _register_artifact_class(self, semantic_type, directory_format,
+                                 description, examples):
+        if not issubclass(directory_format, DirectoryFormat):
+            raise TypeError("%r is not a directory format." % directory_format)
         if not is_semantic_type(semantic_type):
             raise TypeError("%r is not a semantic type." % semantic_type)
-        if not is_semantic_type(semantic_type):
-            raise ValueError("%r is not a semantic type expression."
-                             % semantic_type)
+
         for t in semantic_type:
             if t.predicate is not None:
                 raise ValueError("%r has a predicate, differentiating format"
                                  " on predicate is not supported.")
 
-        self.type_formats.append(TypeFormatRecord(
-            type_expression=semantic_type, format=artifact_format,
-            plugin=self))
+        if description is None:
+            description = ""
+        if examples is None:
+            examples = {}
+
+        registered_artifact_classes = \
+            set(str(e.type_expression) for e in self.artifact_classes)
```

Comment:
This may be worth storing on the plugin instance as a private member, as otherwise we'll be constructing this set for each registration, resulting in n^2 iterations through self.artifact_classes.

A good option might be to turn `types` into a mappingproxy of an internal dict, then you can simply check against `self.types` as things are being constructed.

Reply:
actually instead of `self.types` `self.artifact_classes` would be fine I think.